modules/imgproc/src/contour_approx.cpp

```diff
@@ -0,0 +1,351 @@
+// Removes the points (from the resultant polygon)
+// that have zero 1-curvature
+static vector<ApproxItem> pass_0(const vector<schar> & chain, Point origin, bool isApprox, bool isFull)
+{
+    CV_DbgAssert(chain.size() < (size_t)numeric_limits<int>::max());
+    Point pt = origin;
+    vector<ApproxItem> res;
+    const int len = (int)chain.size();
+    for (int i = 0; i < len; ++i)
+    {
+        const schar prev = (i == 0) ? chain[len - 1] : chain[i - 1];
+        const schar cur = chain[i];
+        const schar s = abs_diff[cur - prev + 7];
+        if ((!isApprox && (isFull || s != 0)) || isApprox)
+        {
+            res.push_back(ApproxItem(pt, s));
+            if (s == 0)
+                (res.end() - 1)->removed = true;
+        }
+        pt += chainCodeDeltas[cur];
+    }
+    return res;
+}
+
+static vector<Point> gatherPoints(const vector<ApproxItem> & ares)
+{
+    vector<Point> res;
+    for (const ApproxItem & item : ares)
+    {
+        if (item.removed)
+            continue;
+        res.push_back(item.pt);
+    }
+    return res;
+}
+
+static int calc_support(const vector<ApproxItem> & ares, int i)
+{
+    int k;
+    int d_num = 0;
+    int l = 0;
+    const int len = (int)ares.size();
+    /* determine support region */
+    for( k = 1;; k++ )
+    {
+        CV_Assert( k <= len );
+        /* calc indices */
+        const int i1 = (i >= k) ? (i - k) : (len - k + i);
+        const int i2 = (i + k < len) ? (i + k) : (i + k - len);
+
+
+        int lk, dk_num;
+        int dx, dy;
+        dx = ares[i2].pt.x - ares[i1].pt.x;
+        dy = ares[i2].pt.y - ares[i1].pt.y;
+
+        /* distance between p_(i - k) and p_(i + k) */
+        lk = dx * dx + dy * dy;
+
+        /* distance between p_i and the line (p_(i-k), p_(i+k)) */
+        dk_num = (ares[i].pt.x - ares[i1].pt.x) * dy - (ares[i].pt.y - ares[i1].pt.y) * dx;
+
+        Cv32suf d;
+        d.f = (float) (((double) d_num) * lk - ((double) dk_num) * l);
+
+        if( k > 1 && (l >= lk || ((d_num > 0 && d.i <= 0) || (d_num < 0 && d.i >= 0))))
+            break;
+
+        d_num = dk_num;
+        l = lk;
+    }
+    return k - 1;
+}
+
+static int calc_cosine(const vector<ApproxItem> & ares, int i)
+{
+    const int k = ares[i].k;
+    int j;
+    int s;
+    const int len = (int)ares.size();
+    /* calc k-cosine curvature */
+    for( j = k, s = 0; j > 0; j-- )
+    {
+        const int i1 = (i >= j) ? (i - j) : (len - j + i);
+        const int i2 = (i + j < len) ? (i + j) : (i + j - len);
+
+        const int dx1 = ares[i1].pt.x - ares[i].pt.x;
+        const int dy1 = ares[i1].pt.y - ares[i].pt.y;
+        const int dx2 = ares[i2].pt.x - ares[i].pt.x;
+        const int dy2 = ares[i2].pt.y - ares[i].pt.y;
+
+        if( (dx1 | dy1) == 0 || (dx2 | dy2) == 0 )
+            break;
+
+        double temp_num = dx1 * dx2 + dy1 * dy2;
+        temp_num =
+            (float) (temp_num /
+                     sqrt( ((double)dx1 * dx1 + (double)dy1 * dy1) *
```

Comment:
Is NaN possible here? E.g. division by zero or sqrt for negative value?

Reply:
Should not be possible, there is a check for 0 (dx, dy are ints):
```.cpp
if( (dx1 | dy1) == 0 || (dx2 | dy2) == 0 )
            break;
```
And sum of squares can not be less than zero too, so sqrt should is also safe.